### PR TITLE
EOS-22149: Hare system tests - Add Bootstrap-shutdown test case to IVT framework

### PR DESCRIPTION
**Pull request by pavankrishnat**
Added Bootstrap-Shutdown test case to IVT framework.

CMD to run Bootstrap-Shutdown test in loop for 10 times.
`/opt/seagate/cortx/hare/bin/hare_setup test --config 'json:///opt/seagate/cortx_configs/provisioner_cluster.json' --file /var/lib/hare/cluster.yaml --plan sanity --dev 10`

Test will run only if `--dev` argument is used.

### CODE REVIEW

File: provisioning/miniprov/hare_mp/main.py
Comment (knekrasov) [inline provisioning/miniprov/hare_mp/main.py:258]: Let it be just `execute`. Cmds is misleading (it executes just one command only).

File: provisioning/miniprov/hare_mp/main.py
Comment (knekrasov) [inline provisioning/miniprov/hare_mp/main.py:267]: The whole method could be looking like this:

```python
    def _execute(cmd: List[str]) -> str:
        process = Popen(cmd,
                        stdin=PIPE,
                        stdout=PIPE,
                        stderr=PIPE,
                        encoding='utf8')
        logging.debug('Issuing CLI command: %s', cmd)
        out, err = process.communicate()
        exit_code = process.returncode
        logging.debug('Finished. Exit code: %d', exit_code)
        if exit_code:
            raise CliException(out, err, exit_code)
        return out
```
So if everything is fine, it will just return the `stdout`. In case of non-zero exit code an exception is thrown (with stderr and exact exit code) - the exception type is custom.

Why do you want to return a list of stdout, stderr and exit code? This result value is not comfortable to handle.

File: provisioning/miniprov/hare_mp/main.py
Comment (knekrasov) [inline provisioning/miniprov/hare_mp/main.py:286]: `resp[2]` looks too magical. Please refer my comment with `_execute` example.

File: provisioning/miniprov/hare_mp/main.py
Comment (knekrasov) [inline provisioning/miniprov/hare_mp/main.py:294]: This deserves a separate function. And please think how to avoid `assert` instruction.

File: provisioning/miniprov/hare_mp/main.py
Comment (knekrasov) [inline provisioning/miniprov/hare_mp/main.py:330]: Sleep-driven development is error-prone.

FYI: in the past I implemented a tool called `pcswrap` that can be helpful for things like this (e.g. start all services and wait until they are started for sure - with timeout). Currently its source code is in cortx-ha repository: https://github.com/Seagate/cortx-ha/tree/main/ha/pcswrap

We could leverage either the whole pcswrap tool (TBD) or reuse some parts of its codebase. For instance, [Waiter](https://github.com/Seagate/cortx-ha/blob/19c6cad22b4a7f2036bc7e292b40383bae4d7fdb/ha/pcswrap/pcswrap/internal/waiter.py#L21) class can be beneficial:

https://github.com/Seagate/cortx-ha/blob/19c6cad22b4a7f2036bc7e292b40383bae4d7fdb/ha/pcswrap/pcswrap/client.py#L97-L103


@mssawant What's your opinion? Can we take back (or copy) pcswrap back to our repository?

File: provisioning/miniprov/hare_mp/main.py
Comment (knekrasov) [inline provisioning/miniprov/hare_mp/main.py:358]: Do you need this decorator? If not, please remove the line. Otherwise please uncomment it,

File: provisioning/miniprov/hare_mp/main.py
Comment (knekrasov) [inline provisioning/miniprov/hare_mp/main.py:393]: Please don't use `exit()`. Especially if this function will become a test runnable by pytest.

File: provisioning/miniprov/hare_mp/main.py
Comment (knekrasov) [inline provisioning/miniprov/hare_mp/main.py:392]:
```suggestion
                logging.info('created journald log: %s', jlog)
```

File: provisioning/miniprov/hare_mp/main.py
Comment (knekrasov) [inline provisioning/miniprov/hare_mp/main.py:402]:
```suggestion
                logging.info('created journald log: %s', jlog)
```

File: provisioning/miniprov/hare_mp/main.py
Comment (knekrasov) [inline provisioning/miniprov/hare_mp/main.py:403]: Please don't use `exit()`

File: provisioning/miniprov/hare_mp/main.py
Comment (knekrasov) [inline provisioning/miniprov/hare_mp/main.py:405]: Why do you need `sleep`?

File: provisioning/miniprov/hare_mp/main.py
Comment (knekrasov) [inline provisioning/miniprov/hare_mp/main.py:753]: What does the number value mean? Is it just a workaround? If so, please consider looking for `store_true` in the [official tutorial](https://docs.python.org/3/howto/argparse.html).

File: provisioning/miniprov/hare_mp/main.py
Comment (knekrasov) [inline provisioning/miniprov/hare_mp/main.py:453]: Here you could have just raised an exception.

File: provisioning/miniprov/hare_mp/main.py
Comment (knekrasov) [inline provisioning/miniprov/hare_mp/main.py:445]: Here you could have just raised an exception.

File: provisioning/miniprov/hare_mp/main.py
Comment (knekrasov) [inline provisioning/miniprov/hare_mp/main.py:441]: Here you could have just raised an exception.

File: provisioning/miniprov/hare_mp/main.py
Comment (knekrasov) [inline provisioning/miniprov/hare_mp/main.py:436]: Log message is repeated both in `if` and in `else` branches. Let's move it out of `if` statment.

File: provisioning/miniprov/hare_mp/main.py
Comment (knekrasov) [inline provisioning/miniprov/hare_mp/main.py:445]: Both if and else branches deserve separate functions. E.g.
```python

if dev_mode:
   ensure_cluster_running()
else:
    run_bootstrap_shutdown_test()
logging.info('Tests executed successfully')
```

File: provisioning/miniprov/hare_mp/main.py
Comment (knekrasov) [inline provisioning/miniprov/hare_mp/main.py:788]: Do you like this stairway of functions?
If not, we could make use of `functools` to improve readability. Example: https://mathieularose.com/function-composition-in-python

**Review by knekrasov — CHANGES_REQUESTED** [review]
Please also squash the commits. Thank you

File: provisioning/miniprov/hare_mp/main.py
Comment (pavankrishnat) [reply provisioning/miniprov/hare_mp/main.py]: used function which is already in use

File: provisioning/miniprov/hare_mp/main.py
Comment (pavankrishnat) [reply provisioning/miniprov/hare_mp/main.py]: For now I will remove the decorator, once PR#1679 merged will move completed code accordingly.

File: provisioning/miniprov/hare_mp/main.py
Comment (pavankrishnat) [reply provisioning/miniprov/hare_mp/main.py]: Sure

File: provisioning/miniprov/hare_mp/main.py
Comment (pavankrishnat) [reply provisioning/miniprov/hare_mp/main.py]: Dev argument is used to separate, my test interface from actual main branch. and using this Dev value as loop count.

File: provisioning/miniprov/hare_mp/main.py
Comment (pavankrishnat) [reply provisioning/miniprov/hare_mp/main.py:742]: I have not created it, I used it, to add my dev argument in IVT framework.

Comment (knekrasov) [reply provisioning/miniprov/hare_mp/main.py:742]: But wouldn't you improve it given that you touched this piece of code anyway?

File: provisioning/miniprov/hare_mp/main.py
Comment (knekrasov) [inline provisioning/miniprov/hare_mp/main.py:310]: the function name tells nothing about what it does and what behavior is expected.

Suggestion: `get_hctl_status()`

File: provisioning/miniprov/hare_mp/main.py
Comment (knekrasov) [inline provisioning/miniprov/hare_mp/main.py:324]: This is a code duplication. Please refactor `execute()` function respectfully instead.

File: provisioning/miniprov/hare_mp/main.py
Comment (pavankrishnat) [reply provisioning/miniprov/hare_mp/main.py:742]: Sure will take it and do this change

File: provisioning/miniprov/hare_mp/main.py
Comment (knekrasov) [inline provisioning/miniprov/hare_mp/main.py:338]: Why do we have tests hardcoded in this main.py file?

Can we have IVT tests be written as valid pytest tests? 
- In this case they would be located in a special directory and/or marked with special [marker](url),
- The only thing  that the code in this mini provisioner CLI would have is to invoke pytest with proper command line arguments

Pros:
1. List of tests become visible and easily extensible
2. There is a standard way to run the tests. As a consequence, such test can be integrated with test coverage tools and with CI (because pytest support standard formats to report the test run results) in the future.


@vaibhavparatwar, @mssawant, @suvratjoshi Any thoughts?

File: provisioning/miniprov/hare_mp/main.py
Comment (knekrasov) [inline provisioning/miniprov/hare_mp/main.py:346]: As far as I understand, this function will just invoke `hctl status` and return the stdout as the return value. That value is ignored. Then what is the point to invoke this function here?

**Review by knekrasov — CHANGES_REQUESTED** [review]

File: provisioning/miniprov/hare_mp/main.py
Comment (pavankrishnat) [reply provisioning/miniprov/hare_mp/main.py:338]: Note: once [PR#1679](https://github.com/Seagate/cortx-hare/pull/1679) is merged I will move my changes accordingly.

File: provisioning/miniprov/hare_mp/main.py
Comment (pavankrishnat) [reply provisioning/miniprov/hare_mp/main.py:324]: hctl status returns non zero error code when cluster is not running, In this case I don't want to exit, so I added few checks.

Comment (knekrasov) [reply provisioning/miniprov/hare_mp/main.py:324]: I don't see a problem here.
1. `execute()` raises an exception if the exit code is non-zero
2. you can catch that exception here in your code (and potentially - analyze the exception).
3. ..and keep processing after that. No exit will be triggered.

But all that unpleasant things like `PIPE`, `returncode` etc would be kept in one place only - in `execute()` function.

File: provisioning/miniprov/hare_mp/main.py
Comment (knekrasov) [reply provisioning/miniprov/hare_mp/main.py:338]: I believe you are not blocked by #1679. In other words, you can start putting ivt tests into some folder in the source code even without that patch. Do I miss anything|?

**Pull request closed without merging**
